### PR TITLE
Script to convert HF checkpoint to T5X

### PR DESCRIPTION
**Pull request by hugoabonizio**
Now we rely on a previously saved T5X checkpoint to overwrite its weights, but we can create one on-the-fly based on HF configs.

```
$ time python3 convert.py --model_name_or_path EleutherAI/gpt-j-6B --t5x_checkpoint_path model/checkpoint_0
2022-11-18 01:02:25.477703: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 AVX512F FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
Remaining (T5X): {'checkpoint'}
Remaining (HF): {'transformer.h.25.attn.masked_bias', 'transformer.h.3.attn.masked_bias', 'transformer.h.16.attn.bias', 'transformer.h.10.attn.bias', 'transformer.h.18.attn.bias', 'transformer.h.14.attn.masked_bias', 'transformer.h.18.attn.masked_bias', 'transformer.h.16.attn.masked_bias', 'transformer.h.19.attn.masked_bias', 'transformer.h.20.attn.masked_bias', 'transformer.h.9.attn.masked_bias', 'transformer.h.13.attn.masked_bias', 'transformer.h.15.attn.bias', 'transformer.h.0.attn.bias', 'transformer.h.9.attn.bias', 'transformer.h.21.attn.masked_bias', 'transformer.h.2.attn.bias', 'transformer.h.19.attn.bias', 'transformer.h.8.attn.bias', 'transformer.h.8.attn.masked_bias', 'transformer.h.26.attn.masked_bias', 'transformer.h.5.attn.masked_bias', 'transformer.h.24.attn.masked_bias', 'transformer.h.4.attn.bias', 'transformer.h.1.attn.masked_bias', 'transformer.h.17.attn.masked_bias', 'transformer.h.14.attn.bias', 'transformer.h.20.attn.bias', 'transformer.h.24.attn.bias', 'transformer.h.7.attn.masked_bias', 'transformer.h.26.attn.bias', 'transformer.h.4.attn.masked_bias', 'transformer.h.17.attn.bias', 'transformer.h.27.attn.masked_bias', 'transformer.h.0.attn.masked_bias', 'transformer.h.27.attn.bias', 'transformer.h.11.attn.bias', 'transformer.h.25.attn.bias', 'transformer.h.7.attn.bias', 'transformer.h.11.attn.masked_bias', 'transformer.h.6.attn.masked_bias', 'transformer.h.6.attn.bias', 'transformer.h.1.attn.bias', 'transformer.h.23.attn.bias', 'transformer.h.5.attn.bias', 'transformer.h.22.attn.bias', 'transformer.h.2.attn.masked_bias', 'transformer.h.12.attn.bias', 'transformer.h.22.attn.masked_bias', 'transformer.h.15.attn.masked_bias', 'transformer.h.13.attn.bias', 'transformer.h.23.attn.masked_bias', 'transformer.h.3.attn.bias', 'transformer.h.21.attn.bias', 'transformer.h.12.attn.masked_bias', 'transformer.h.10.attn.masked_bias'}
Done!

real	16m9.656s
user	20m57.417s
sys	1m5.963s
```

- [x] Validate converted weights
- [ ] Create checkpoint dynamically
- [ ] Optimize (?)

### CODE REVIEW

File: t5x/scripts/convert_hf_checkpoint.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        "--model_name_or_path", type=str, required=True, help="HuggingFace model."
+    )
+    parser.add_argument(
+        "--t5x_checkpoint_path", type=str, required=True, help="T5x checkpoint to be overwritten."
+    )
+    args = parser.parse_args()
+    convert_hf_checkpoint_to_t5x(args.model_name_or_path, args.t5x_checkpoint_path)
+
+# TODO
+# understand what are transformer.h.0.attn.bias and transformer.h.0.attn.masked_bias
```
Comment (rodrigonogueira4) [inline t5x/scripts/convert_hf_checkpoint.py:82]: is this something in the HF checkpoints?

File: t5x/scripts/convert_hf_checkpoint.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
+import os
+import argparse
+import tensorstore as ts
+from transformers import AutoModelForCausalLM, AutoConfig
+
+# from t5x.checkpoints import load_t5x_checkpoint
```
Comment (rodrigonogueira4) [inline t5x/scripts/convert_hf_checkpoint.py:6]: Maybe add a sentence above saying "Example usage:"?

File: t5x/scripts/convert_hf_checkpoint.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+# ckpt = load_t5x_checkpoint('gs://rodrigo-temp3/t5x/hugo_test1/checkpoint_0')
```
Comment (rodrigonogueira4) [inline t5x/scripts/convert_hf_checkpoint.py:7]: Use more "generic" paths?

File: t5x/scripts/convert_hf_checkpoint.py (hunk 1 above)
Comment (hugoabonizio) [reply t5x/scripts/convert_hf_checkpoint.py:82]: Yes, they're defined here and saved along with the model weights while in T5X they're created on-the-fly
HF: https://github.com/huggingface/transformers/blob/94b3f544a1f5e04b78d87a2ae32a7ac252e22e31/src/transformers/models/gptj/modeling_gptj.py#L98-L104
T5X: https://github.com/rodrigonogueira4/t5x/blob/main/t5x/examples/decoder_only/layers.py#L468-L473

File: t5x/scripts/convert_hf_checkpoint.py (hunk 2 above)
Comment (hugoabonizio) [reply t5x/scripts/convert_hf_checkpoint.py:6]: These are actually leftovers. I was thinking of using this method to double-check the conversion by loading the weights and comparing again.

File: t5x/scripts/convert_hf_checkpoint.py (hunk 3 above)
Comment (hugoabonizio) [reply t5x/scripts/convert_hf_checkpoint.py:7]: These lines will be gone by the final version!

Comment (rodrigonogueira4) [reply t5x/scripts/convert_hf_checkpoint.py:7]: ah ok, thanks!

File: t5x/scripts/convert_hf_checkpoint.py (hunk 1 above)
Comment (rodrigonogueira4) [reply t5x/scripts/convert_hf_checkpoint.py:82]: An this is only for the first layer (h.0) or the others as well?

Comment (hugoabonizio) [reply t5x/scripts/convert_hf_checkpoint.py:82]: For all attention layers. I'm priting the weights that were not converted in the end of the script:
```
Remaining (HF): {'transformer.h.25.attn.masked_bias', 'transformer.h.3.attn.masked_bias', 'transformer.h.16.attn.bias' (...)
```

**Comment by ZanezZephyrs** [comment]
About validating the converted weights.
I converted the t5x checkpoint back to HF format, and then compared the weights with the original gptj. They were a 100% match.
In order to convert the checkpoint back to HF i had to copy the optimizer states from another checkpoint, but these are not important for the conversion.

Code i used to check

`original_model = GPTJForCausalLM.from_pretrained(
    "EleutherAI/gpt-j-6B")

candidate_model = GPTJForCausalLM.from_pretrained(
    "output/converted")

for (original_name,original_param),(candidate_name,candidate_param) in zip(original_model.named_parameters(),candidate_model.named_parameters()):
    print(original_name,candidate_name)
    print(torch.sum(torch.eq(torch.tensor(original_param),torch.tensor(candidate_param))).item()/torch.tensor(original_param).nelement())`

**Comment by hugoabonizio** [comment]
I'll leave the creation of dynamic checkpoints and possible optimizations for later, in case they become necessary.

**Review by rodrigonogueira4 — APPROVED** [review]
Thanks a lot for implementing this script!

**Pull request merged**